File: matmult.f90
! matmult.f90 - matrix multiplication if Fortran 90

PROGRAM matmult

   INTEGER i, j, k, N, cs, ce, cr
   DOUBLE PRECISION :: t_mult
   DOUBLE PRECISION, DIMENSION(1000,1000) :: A, B, C

      ! Allocate matrices and initialize A and B

   N = 1000
   DO i = 1, N
      DO j = 1, N
         A( i, j ) = DBLE( i + j )
         B( i, j ) = DBLE( i * j + 1.0)
      END DO
   END DO

      ! Now time the multiplication

   CALL SYSTEM_CLOCK( COUNT_RATE = cr )
   CALL SYSTEM_CLOCK( COUNT = cs )

   DO i = 1, N
      DO j = 1, N
         C(i,j) = 0.0
         DO k = 1, N
            C( i, j ) = C( i, j ) + A( i, k ) * B( k, j )
         END DO
      END DO
   END DO

   CALL SYSTEM_CLOCK( COUNT = ce )
   t_mult = DBLE( ce - cs ) / cr

   WRITE(*,*) "Multiplying 2 matrices of size ", N
   WRITE(*,*) "Time ", t_mult, " seconds"

END PROGRAM matmult

File: matmult.m
% Matrix multiplication in Matlab using serial for, parfor, and builtin methods
% The builtin parallel routine will use all cores available


   % Allocate and initialize the A and B matrices

N=100;
ncores=4;

A=rand( N, N );
B=rand( N, N );


   % Time a serial matrix multiply

C=zeros( N, N, 'double' );
tic  % start the timer

for i = 1:N
   for j = 1:N
      for k = 1:N
         C(i,j) = C(i,j) + A(i,k)*B(k,j);
      end
   end
end

t_end = toc;
%disp(C);
disp(['serial loop took ', num2str(t_end), ' seconds']);


   % Time the internal matrix multiply which is multi-threaded

C=zeros( N, N, 'double' );
tic  % start the timer

C = A * B;

t_end = toc;
%disp(C);
disp(['builtin matmult took ', num2str(t_end), ' seconds']);


   % Time a single-node multi-process  parfor matrix multiply

tic

parpool('local', ncores)    % Fire up the virtual cluster of processes

t_end = toc;
disp([num2str(t_end), ' seconds to spin up multi-process pool']);

C=zeros( N, N, 'double' );
tic  % start the timer

parfor i = 1:N
   for j = 1:N
      for k = 1:N
         C(i,j) = C(i,j) + A(i,k)*B(k,j);
      end
   end
end

t_end = toc;
%disp(C);
disp(['multi-process parfor loop took ', num2str(t_end), ' seconds']);


exit; % NOTE: Multi-threads not supported in older versions of Matlab


   % Time a single-node multi-thread parfor matrix multiply

delete( gcp( 'nocreate' ) );   % Destroy the local parpool

tic

parpool('threads', ncores)

t_end = toc;
disp([num2str(t_end), ' seconds to spin up multi-thread pool']);

C=zeros( N, N, 'double' );
tic  % start the timer

parfor i = 1:N
   for j = 1:N
      for k = 1:N
         C(i,j) = C(i,j) + A(i,k)*B(k,j);
      end
   end
end

t_end = toc;
%disp(C);
disp(['multi-threaded parfor loop took ', num2str(t_end), ' seconds']);

exit;

